src/base/client.zig

```diff
@@ -232,15 +235,20 @@ pub fn BaseClient(comptime Reader: type, comptime Writer: type) type {
             }
         }
 
+        pub fn closeReader(self: *Self) void {
+            self.self_contained = false;
+            self.payload_size = 0;
+        }
+
         pub fn readNextChunkBuffer(self: *Self, buffer: []u8) ReadNextError!usize {
-            if (self.parser.state != .chunk) {
-                self.self_contained = false;
-                return 0;
-            }
+            if (self.payload_index >= self.payload_size) {
+                if (self.parser.state != .chunk) {
```

Comment:
I moved this chunk test here because otherwise this triggers immediately — the final chunk resets the state so the  reader closes way too early.